# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmBaselineCompile.cpp
# Commit: 1ab8383d17ee
# Full Hash: 1ab8383d17ee1e5db164339236d64bcb7168b225
# Author: Lars T Hansen <lhansen@mozilla.com>
# Date: 2021-10-18 16:09:29
# Regressor Bug: 1727084
# File Overlap Count: 1
# Description:
#   Bug 1727084 - Memory64 - Atomic-rmw baseline operations. r=yury
#   
#   Atomic ADD, SUB, AND, OR, XOR, XCHG, CMPXCHG, as well as WAIT and
#   NOTIFY for the baseline compiler.
#   
# ==============================================================================

diff -r 6089b4360edc -r 1ab8383d17ee js/src/wasm/WasmBaselineCompile.cpp
--- a/js/src/wasm/WasmBaselineCompile.cpp	Mon Oct 18 07:31:55 2021 +0000
+++ b/js/src/wasm/WasmBaselineCompile.cpp	Mon Oct 18 07:31:56 2021 +0000
@@ -291,6 +291,31 @@
 #endif
 }
 
+#ifdef JS_CODEGEN_X86
+void BaseCompiler::stashI64(RegPtr regForTls, RegI64 r) {
+  MOZ_ASSERT(sizeof(TlsData::baselineScratch) >= 8);
+  MOZ_ASSERT(regForTls != r.low && regForTls != r.high);
+  fr.loadTlsPtr(regForTls);
+  masm.store32(r.low, Address(regForTls, offsetof(TlsData, baselineScratch)));
+  masm.store32(r.high,
+               Address(regForTls, offsetof(TlsData, baselineScratch) + 4));
+}
+
+void BaseCompiler::unstashI64(RegPtr regForTls, RegI64 r) {
+  MOZ_ASSERT(sizeof(TlsData::baselineScratch) >= 8);
+  fr.loadTlsPtr(regForTls);
+  if (regForTls == r.low) {
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch) + 4),
+                r.high);
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch)), r.low);
+  } else {
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch)), r.low);
+    masm.load32(Address(regForTls, offsetof(TlsData, baselineScratch) + 4),
+                r.high);
+  }
+}
+#endif
+
 //////////////////////////////////////////////////////////////////////////////
 //
 // Function entry and exit